auth/user_mgt.go

```diff
@@ -0,0 +1,493 @@
+	PhotoURL           string      `json:"photoUrl,omitempty"`
+	CreationTimestamp  int64       `json:"createdAt,string,omitempty"`
+	LastLogInTimestamp int64       `json:"lastLoginAt,string,omitempty"`
+	ProviderID         string      `json:"providerId,omitempty"`
+	CustomClaims       string      `json:"customAttributes,omitempty"` // https://play.golang.org/p/JB1_jHu1mm
+	Disabled           bool        `json:"disabled,omitempty"`
+	EmailVerified      bool        `json:"emailVerified,omitempty"`
+	ProviderUserInfo   []*UserInfo `json:"providerUserInfo,omitempty"`
+	PasswordHash       string      `json:"passwordHash,omitempty"`
+	PasswordSalt       string      `json:"salt,omitempty"`
+	ValidSince         int64       `json:"validSince,string,omitempty"`
+}
+
+type listUsersResponse struct {
+	RequestType string               `json:"kind,omitempty"`
+	Users       []responseUserRecord `json:"users,omitempty"`
+	NextPage    string               `json:"nextPageToken,omitempty"`
+}
+
+func makeExportedUser(rur responseUserRecord) (*ExportedUserRecord, error) {
+	var ccp *map[string]interface{}
+	if rur.CustomClaims != "" {
+		var cc map[string]interface{}
+		err := json.Unmarshal([]byte(rur.CustomClaims), &cc)
+		if err != nil {
+			return nil, err
+		}
+		if len(cc) > 0 {
+			ccp = &cc
+		}
+	}
+
+	resp := &ExportedUserRecord{
+		UserRecord: &UserRecord{
+			UserInfo: &UserInfo{
+				DisplayName: rur.DisplayName,
+				Email:       rur.Email,
+				PhoneNumber: rur.PhoneNumber,
+				PhotoURL:    rur.PhotoURL,
+				ProviderID:  rur.ProviderID,
+				UID:         rur.UID,
+			},
+			CustomClaims:     ccp,
+			Disabled:         rur.Disabled,
+			EmailVerified:    rur.EmailVerified,
+			ProviderUserInfo: rur.ProviderUserInfo,
+			UserMetadata: &UserMetadata{
+				LastLogInTimestamp: rur.LastLogInTimestamp,
+				CreationTimestamp:  rur.CreationTimestamp,
+			},
+		},
+		PasswordHash: rur.PasswordHash,
+		PasswordSalt: rur.PasswordSalt,
+	}
+	return resp, nil
+}
+
+func (it *UserIterator) fetch(pageSize int, pageToken string) (string, error) {
+	params := map[string]interface{}{"maxResults": pageSize}
+	if pageToken != "" {
+		params["nextPageToken"] = pageToken
+	}
+	resp, err := it.client.makeUserRequest(it.ctx, "downloadAccount", params)
+	if err != nil {
+		// remove this line before submission ,see b/69406469
+		if pageToken != "" &&
+			strings.Contains(err.Error(), "\"code\": 400") &&
+			strings.Contains(err.Error(), "\"message\": \"INVALID_PAGE_SELECTION\"") {
+			return it.fetch(pageSize, "")
+		}
+		return "", err
+	}
+	var lur listUsersResponse
+	err = json.Unmarshal(resp, &lur)
+	if err != nil {
+		return "", err
+	}
+	for _, u := range lur.Users {
+		eu, err := makeExportedUser(u)
+		if err != nil {
+			return "", err
+		}
+		it.users = append(it.users, eu)
+	}
+	it.pageInfo.Token = lur.NextPage
+	return lur.NextPage, nil
+}
+
+func validateString(s *string, condition func(string) bool, message string) *string {
```

Comment:
Recommend removing these 3 helpers, and having the respective logic in the caller functions. That seems more readable.

Reply:
I kept one to alleviate the !=nil duplication, 
did you have something else in mind?